File: R/fdrAdjustment.R
fdrAdjustment <- function(data, ranks=NULL, method=c("step.up", "step.down"), c=NULL) {
    method <- match.arg(method)
    stepUp <- method == "step.up"
    if (is.null(c))
        c <- log(length(data)) - digamma(1)
    if (missing(ranks))
        ranks <- rank(data)
    q <- (data * length(data) * c / ranks)[order(data)]
    q <- .C("_fdrAdjustment",
        q = as.double(q),
        as.integer(length(q)),
        as.integer(stepUp))$q[ranks]
    data.frame(rank=ranks, q)
}

File: src/fdrAdjustment.c
#include <math.h>

void fdrAdjust(double *q, int *n, int *stepUp) {
    int i;
    double p = 1;

    if (*stepUp) {
        for (i = *n - 1; i >= 0; i--) {
            if (q[i] < p) p = q[i];
            q[i] = p;
        }
    } else {
        for (i = 0; i < *n; i++) {
            if (q[i] > p) p = q[i];
            q[i] = p;
        }
    }
}
